crates/api/src/transport.rs

```diff
@@ -0,0 +1,350 @@
+    pub async fn send_module(
+        &self,
+        peer: Url,
+        space: SpaceId,
+        module: String,
+        data: bytes::Bytes,
+    ) -> K2Result<()> {
+        let enc = (K2Proto {
+            ty: k2_proto::Ty::Module as i32,
+            data,
+            space: Some(space.into()),
+            module: Some(module),
+        })
+        .encode()?;
+        self.imp.send(peer, enc).await
+    }
+}
+
+/// Base trait for transport handler events.
+/// The other three handler types are all based on this trait.
+pub trait TxBaseHandler: 'static + Send + Sync + std::fmt::Debug {
+    /// A notification that a new listening address has been bound.
+    /// Peers should now go to this new address to reach this node.
+    fn new_listening_address(&self, this_url: Url) {
+        drop(this_url);
+    }
+
+    /// A peer has connected to us. In addition to the preflight
+    /// logic in [TxHandler], this callback allows space and module
+    /// logic to block connections to peers. Simply return an Err here.
+    fn peer_connect(&self, peer: Url) -> K2Result<()> {
+        drop(peer);
+        Ok(())
+    }
+
+    /// A peer has disconnected from us. If they did so gracefully
+    /// the reason will be is_some().
+    fn peer_disconnect(&self, peer: Url, reason: Option<String>) {
+        drop((peer, reason));
+    }
+}
+
+/// Handler for whole transport-level events.
+pub trait TxHandler: TxBaseHandler {
+    /// Gather preflight data to send to a new opening connection.
+    /// Returning an Err result will close this connection.
+    ///
+    /// The default implementation sends an empty preflight message.
+    fn preflight_gather_outgoing(
+        &self,
+        peer_url: Url,
+    ) -> K2Result<bytes::Bytes> {
+        drop(peer_url);
+        Ok(bytes::Bytes::new())
+    }
+
+    /// Validate preflight data sent by a remote peer on a new connection.
+    /// Returning an Err result will close this connection.
+    ///
+    /// The default implementation ignores the preflight data,
+    /// and considers it valid.
+    fn preflight_validate_incoming(
+        &self,
+        peer_url: Url,
+        data: bytes::Bytes,
+    ) -> K2Result<()> {
+        drop((peer_url, data));
+        Ok(())
+    }
+}
+
+/// Trait-object [TxHandler].
+pub type DynTxHandler = Arc<dyn TxHandler>;
+
+/// Handler for space-related events.
+pub trait TxSpaceHandler: TxBaseHandler {
+    /// The sync handler for receiving notifications sent by a remote
+    /// peer in reference to a particular space.
+    fn recv_space_notify(&self, peer: Url, space: SpaceId, data: bytes::Bytes) {
+        drop((peer, space, data));
+    }
+}
+
+/// Trait-object [TxSpaceHandler].
+pub type DynTxSpaceHandler = Arc<dyn TxSpaceHandler>;
+
+/// Handler for module-related events.
+pub trait TxModuleHandler: TxBaseHandler {
+    /// The sync handler for receiving module messages sent by a remote
+    /// peer in reference to a particular space.
+    fn recv_module(
```

Comment:
I wonder if this should be `recv_module_msg`. It's the message that gets received and not the module. I'm trying to get used to this but I keep tripping up on the meaning of this

Reply:
haha, fair : )

Reply:
Done: https://github.com/holochain/kitsune2/pull/45/commits/f5bf55b6564d64f4e2144fe2ec52e7fdb7f791de